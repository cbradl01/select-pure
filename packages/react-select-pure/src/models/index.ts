export * from "./select-pure.models";
export * from "./option-pure.models";
export * from "./select.models";
export * from "./option.models";
