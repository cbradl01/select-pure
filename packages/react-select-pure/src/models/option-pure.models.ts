export interface OptionPureProps {
  children?: any;
  value: string;
  label: string;
  hidden?: boolean;
  selected?: boolean;
  disabled?: boolean;
}
