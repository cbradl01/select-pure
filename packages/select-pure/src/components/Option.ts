import { boundMethod } from "autobind-decorator";
import { LitElement, html } from "lit";
import { ifDefined } from "lit-html/directives/if-defined.js";
import { customElement } from "lit/decorators/custom-element.js";
import { property } from "lit/decorators/property.js";

import { KEYS } from "./../constants";
import { optionStyles } from "./../styles";

@customElement("option-pure")
export class OptionPure extends LitElement {
  static get styles() {
    return optionStyles;
  }

  @property() isSelected: boolean = false;
  @property() isDisabled: boolean = false;
  @property() isHidden: boolean = false;
  @property() optionValue: string = "";
  @property() displayedLabel: string = "";
  @property() optionIndex: number = -1;

  private onSelect?: Function;
  private onReady?: Function;

  connectedCallback() {
    super.connectedCallback();

    this.isSelected = this.getAttribute("selected") !== null;
    this.isDisabled = this.getAttribute("disabled") !== null;
    this.isHidden = this.getAttribute("hidden") !== null;
    this.optionValue = this.getAttribute("value") || "";
    this.assignDisplayedLabel();

    this.fireOnReadyCallback();
  }

  @boundMethod
  public getOption() {
    return {
      label: this.displayedLabel,
      value: this.optionValue,
      select: this.select,
      unselect: this.unselect,
      selected: this.isSelected,
      disabled: this.isDisabled,
      hidden: this.isHidden,
    };
  }

  @boundMethod
  public select() {
    this.isSelected = true;
    this.setAttribute("selected", "");
  }

  @boundMethod
  public unselect() {
    this.isSelected = false;
    this.removeAttribute("selected");
  }

  public setOnReadyCallback(onReadyCallback: Function, optionIndex: number) {
    this.onReady = onReadyCallback;
    this.optionIndex = optionIndex;
  }

  public setOnSelectCallback(callback: Function) {
    this.onSelect = callback;
  }

  render() {
    const optionWrapperClassNames = ["option"];
    if (this.isSelected) {
      optionWrapperClassNames.push("selected");
    }
    if (this.isDisabled) {
      optionWrapperClassNames.push("disabled");
    }
    return html`
      <div
        class="${optionWrapperClassNames.join(" ")}"
        @click=${this.fireOnSelectCallback}
        @keydown="${this.fireOnSelectIfEnterPressed}"
        tabindex="${ifDefined(this.isDisabled ? undefined : "0")}"
      >
        <slot hidden @slotchange=${this.assignDisplayedLabel}></slot>
        <p>
        ${this.displayedLabel}
        <span></span>
        </p>
      </div>
    `;
  }

  private assignDisplayedLabel() {
    if (this.textContent) {
      this.displayedLabel = this.textContent;
      return;
    }
    if (this.getAttribute("label")) {
      this.displayedLabel = this.getAttribute("label") || "";
    }
  }

  @boundMethod
  private fireOnReadyCallback() {
    if (!this.onReady) {
      return;
    }
    this.onReady(this.getOption(), this.optionIndex);
  }

  private fireOnSelectCallback(optionClickEvent: Event) {
    optionClickEvent.stopPropagation();
    if (!this.onSelect || this.isDisabled) {
      return;
    }
    this.onSelect(this.optionValue);
  }

  private fireOnSelectIfEnterPressed(keyboardKeydownEvent: KeyboardEvent) {
    if (keyboardKeydownEvent.key === KEYS.ENTER) {
      this.fireOnSelectCallback(keyboardKeydownEvent);
    }
  }
}
