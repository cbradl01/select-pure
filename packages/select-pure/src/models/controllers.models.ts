import { TemplateResult } from "lit";

export interface SelectController {
  // getDisplayedLabel: () => TemplateResult;
  // getValue: () => string | string[];
  selectedIndex: number;
}
