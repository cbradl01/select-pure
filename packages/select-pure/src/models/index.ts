export * from "./option.models";
export * from "./controllers.models";
