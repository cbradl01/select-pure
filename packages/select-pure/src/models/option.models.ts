import {LitElement} from "lit";

export interface Option {
  label: string;
  value: string;
  select: Function;
  unselect: Function;
  hidden: boolean;
  disabled: boolean;
}

export interface OptionPureElement extends LitElement {
  getOption: Function;
  setOnSelectCallback: Function;
}
