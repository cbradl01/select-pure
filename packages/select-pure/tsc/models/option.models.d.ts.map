{"version":3,"file":"option.models.d.ts","sourceRoot":"","sources":["../../src/models/option.models.ts"],"names":[],"mappings":"AAAA,OAAO,EAAC,UAAU,EAAC,MAAM,KAAK,CAAC;AAE/B,MAAM,WAAW,MAAM;IACrB,KAAK,EAAE,MAAM,CAAC;IACd,KAAK,EAAE,MAAM,CAAC;IACd,MAAM,EAAE,QAAQ,CAAC;IACjB,QAAQ,EAAE,QAAQ,CAAC;IACnB,MAAM,EAAE,OAAO,CAAC;IAChB,QAAQ,EAAE,OAAO,CAAC;CACnB;AAED,MAAM,WAAW,iBAAkB,SAAQ,UAAU;IACnD,SAAS,EAAE,QAAQ,CAAC;IACpB,mBAAmB,EAAE,QAAQ,CAAC;CAC/B"}