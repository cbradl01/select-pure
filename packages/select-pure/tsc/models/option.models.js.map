{"version":3,"file":"option.models.js","sourceRoot":"","sources":["../../src/models/option.models.ts"],"names":[],"mappings":"","sourcesContent":["import {LitElement} from \"lit\";\n\nexport interface Option {\n  label: string;\n  value: string;\n  select: Function;\n  unselect: Function;\n  hidden: boolean;\n  disabled: boolean;\n}\n\nexport interface OptionPureElement extends LitElement {\n  getOption: Function;\n  setOnSelectCallback: Function;\n}\n"]}