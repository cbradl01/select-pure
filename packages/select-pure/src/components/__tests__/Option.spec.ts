import "./../Option";
import { OptionPureElement } from "./../../models";

describe("Option component", () => {
  it("renders option", async() => {
    document.body.innerHTML = "<option-pure value='UA'>Ukraine</option-pure>";
    const option = document.body.querySelector("option-pure") as OptionPureElement;

    await option.updateComplete;

    expect(option.shadowRoot?.querySelector(".option")?.textContent?.trim()).toEqual("Ukraine");
    expect(option.shadowRoot?.querySelector(".option")?.getAttribute("class")).toEqual("option");
  });

  it("renders with label as an attribute", async() => {
    document.body.innerHTML = "<option-pure value='PL' label='Poland'></option-pure>";
    const option = document.body.querySelector("option-pure") as OptionPureElement;

    await option.updateComplete;

    expect(option.shadowRoot?.querySelector(".option")?.textContent?.trim()).toEqual("Poland");
    expect(option.shadowRoot?.querySelector(".option")?.getAttribute("class")).toEqual("option");
  });

  it("fires onSelect callback", async() => {
    document.body.innerHTML = "<option-pure value='PL' label='Poland'></option-pure>";
    const option = document.body.querySelector("option-pure") as OptionPureElement;
    const mockedOnSelect = jest.fn();

    option.setOnSelectCallback(mockedOnSelect);

    await option.updateComplete;

    const optionDiv = option.shadowRoot?.querySelector(".option");

    optionDiv?.dispatchEvent(new Event("click"));

    expect(mockedOnSelect).toHaveBeenCalledTimes(1);
    expect(mockedOnSelect).toHaveBeenCalledWith("PL");
  });

  it("renders disabled", async() => {
    document.body.innerHTML = "<option-pure value='PL' label='Poland' disabled></option-pure>";
    const option = document.body.querySelector("option-pure") as OptionPureElement;
    const mockedOnSelect = jest.fn();

    option.setOnSelectCallback(mockedOnSelect);

    await option.updateComplete;

    expect(option.shadowRoot?.querySelector(".option")?.getAttribute("class")).toEqual("option disabled");

    const optionDiv = option.shadowRoot?.querySelector(".option");

    optionDiv?.dispatchEvent(new Event("click"));

    expect(mockedOnSelect).not.toHaveBeenCalled();
  });

  // eslint-disable-next-line
  it("renders selected", () => {});
  // eslint-disable-next-line
  it("renders selected", () => {});

  it("fires onReady callback", async() => {
    document.body.innerHTML = "<option-pure value='PL'>Poland</option-pure>";
    const option = document.body.querySelector("option-pure") as OptionPureElement;

    const mockedOnReady = jest.fn();
    // eslint-disable-next-line no-magic-numbers
    option.setOnReadyCallback(mockedOnReady, 177);

    await option.updateComplete;
    // :(
    option.connectedCallback();

    expect(mockedOnReady).toHaveBeenCalledTimes(1);
    expect(mockedOnReady.mock.calls).toMatchSnapshot();
    expect(option.shadowRoot?.querySelector(".option")?.textContent?.trim()).toEqual("Poland");
    expect(option.shadowRoot?.querySelector(".option")?.getAttribute("class")).toEqual("option");
  });
});
