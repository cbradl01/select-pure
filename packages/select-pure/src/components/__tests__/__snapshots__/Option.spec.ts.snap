// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option component fires onReady callback 1`] = `
Array [
  Array [
    Object {
      "disabled": false,
      "label": "Poland",
      "select": [Function],
      "selected": false,
      "unselect": [Function],
      "value": "PL",
    },
    177,
  ],
]
`;
